Escape semicolons in exported CMake variables

drbenmorgan opened the pull request:
Noted in ATLAS builds that the export of `CMAKE_CUDA_ARCHITECTURES` to the `Config.hh` and `CeleritasConfig.cmake` files resulted in newlines replacing the semicolons when this variable is a proper CMake list. For example, running CMake as

```
$ cmake -DCMAKE_CUDA_ARCHITECTURES="50;61;72;80" ...
```

as in ATLAS builds created a `Config.hh` with:

```c++
inline constexpr char celeritas_gpu_architectures[] = "50
61
75
80";
```

and thus a compile-time error. Similarly, `CeleritasConfig.cmake` got

```cmake
set(CELERITAS_CMAKE_CUDA_ARCHITECTURES "50
61
75
80")
```

Assuming that we want to retain these as `"X;Y;Z"`, the changes here explicitly escape any semicolons in variables exported to CMake files to ensure their retention, and also the CUDA/HIP architecture variable exported to `Config.hh` to retain them and prevent compilation error.

drbenmorgan committed:
Escape semicolons in exported CMake variables

Noted in ATLAS builds that export of CMAKE_CUDA_ARCHITECTURES to
config header and CMake script results does not handle the case where
this is a ;-separated CMake list. Export of this data to the header
and script results in the semicolons being replaced by newlines and
consequently a compile error will result. Triaged to exported CMake
variables not having possible semicolons escaped to retain them.

Explicitly escape any semicolons in variables exported to CMake files
to ensure their retention. Escape semicolons in CUDA/HIP architecture
variable exported to Config.hh to retain them and prevent compilation
error.

sethrj reviewed (approved):
Ugh, good catch @drbenmorgan . As a guard against invalid C++ strings we could also wrap them with `R"({var})"` in case they have quotes or whatever else as well.

See https://gitlab.kitware.com/cmake/cmake/-/issues/23315 .

drbenmorgan committed:
Guard exported CMake strings/lists as raw string literals

Provides additional protection against invalid C++ strings that may
result from CMake list/string oddities.
